digdag-core/src/main/java/io/digdag/core/database/DatabaseTaskQueueServer.java

```diff
@@ -0,0 +1,494 @@
+    {
+        ImmutableList.Builder<String> notFoundList = ImmutableList.builder();
+        for (String formatted : lockedIds) {
+            boolean success;
+            if (isSharedTaskLockId(formatted)) {
+                success = taskHeartbeat0(siteId, null, parseTaskLockId(formatted), agentId, lockSeconds);
+            }
+            else {
+                success = taskHeartbeat0(siteId, parseQueueId(formatted), parseTaskLockId(formatted), agentId, lockSeconds);
+            }
+            if (!success) {
+                notFoundList.add(formatted);
+            }
+        }
+        return notFoundList.build();
+    }
+
+    private boolean taskHeartbeat0(int siteId, Integer queueId, long taskLockId, String agentId, int lockSeconds)
+    {
+        return autoCommit((handle, dao) -> {
+            String lockExpireTimeSql;
+            if (isEmbededDatabase()) {
+                lockExpireTimeSql = Long.toString(Instant.now().getEpochSecond() + lockSeconds);
+            }
+            else {
+                lockExpireTimeSql = statementUnixTimestampSql() + " + " + Integer.toString(lockSeconds);
+            }
+            return handle.createStatement(
+                    "update queued_task_locks" +
+                    " set lock_expire_time = " + lockExpireTimeSql +
+                    " where id = :id" +
+                    " and lock_agent_id = :agentId" +
+                    " and coalesce(site_id, (select site_id from queue_settings where id = :queueId)) = :siteId"
+                )
+                .bind("expireTime", Instant.now().getEpochSecond() + lockSeconds)
+                .bind("id", taskLockId)
+                .bind("agentId", agentId)
+                .bind("queueId", queueId)
+                .bind("siteId", siteId)
+                .execute();
+        }) > 0;
+    }
+
+    @Override
+    public List<TaskRequest> lockSharedAgentTasks(int count, String agentId, int lockSeconds, long maxSleepMillis)
+    {
+        for (int siteId : autoCommit((handle, dao) -> dao.getActiveSiteIdList())) {
+            List<Long> taskLockIds = tryLockSharedAgentTasks(siteId, count, agentId, lockSeconds);
+            if (!taskLockIds.isEmpty()) {
+                ImmutableList.Builder<TaskRequest> builder = ImmutableList.builder();
+                for (long taskLockId : taskLockIds) {
+                    byte[] data = autoCommit((handle, dao) -> dao.getTaskData(taskLockId));
+                    if (data == null) {
+                        // queued_task is deleted after tryLockSharedAgentTasks call.
+                        // it is possible just because there are 2 different transactions.
+                    }
+                    else {
+                        String lockId = formatSharedTaskLockId(taskLockId);
+                        builder.add(decodeLockedTaskObject(data, lockId));
+                    }
+                }
+                return builder.build();
```

Comment:
> // it is possible just because there are 2 different transactions.

If it can often happen, how about trying next site id like this?
```
                List<TaskRequest> taskRequests = builder.build();
                if (taskRequests.isEmpty()) {
                    continue;
                }
                return taskRequests;
```